src/prefect/concurrency/services.py

```diff
@@ -127,16 +139,37 @@ async def acquire(
             httpx.HTTPStatusError: If the server returns an error other than 423 LOCKED
             TimeoutError: If acquisition times out
         """
+        use_cache = _should_use_cache(self.concurrency_limit_names, holder)
+        cache_key = frozenset(self.concurrency_limit_names)
+
         with timeout_async(seconds=timeout_seconds):
             while True:
                 try:
-                    return await self._client.increment_concurrency_slots_with_lease(
-                        names=self.concurrency_limit_names,
-                        slots=slots,
-                        mode=mode,
-                        lease_duration=lease_duration,
-                        holder=holder,
+                    if use_cache:
+                        with _cache_lock:
+                            if _no_limits_cache.get(cache_key, False):
+                                return _create_empty_limits_response()
```

Comment:
is there anything that can happen in this `with` scope if the clause is false?

Reply:
my point is that you're eagerly acquiring the lock before you know if you need to, no?